Preparing auto-clustering

dirkmoors opened the pull request:
In this PR I've added a script from http://chad.ill.ac/post/35967173942/easy-ejabberd-clustering-guide-mnesia-mysql and https://github.com/chadillac/ejabberd-easy_cluster which makes it easier to setup a multi master cluster.

Also, I've made SSL optional for s2s and webadmin. For webadmin it is desirable to have disabled SSL on the ejabberd node because you might want to offload SSL to the load balancer (e.g. Haproxy)

rroemhild commented:
That looks great. `erlc` can change the output dir with the `-o` switch, so we do not need to switch to root.

```
# Compile erl scripts
RUN erlc -W0 -o $EJABBERD_ROOT/lib/stdlib/ebin \
    $EJABBERD_ROOT/bin/scripts/erl/*.erl
```

rroemhild commented:
`$EJABBERD_ROOT/lib/stdlib/ebin` is not a good place for ejabberd-modules maybe the ejabberd bin directory described in https://www.ejabberd.im/ejabberd-modules is a better place for the beam files.

```
# Compile erl scripts
RUN erlc -W0 -o $EJABBERD_ROOT/bin \
    $EJABBERD_ROOT/bin/scripts/erl/*.erl
```

dirkmoors commented:
I've updated looked into it and it seems that the current way to go is to have a variable 'CONTRIB_MODULES_PATH' in ejabberdct.cfg. (see: https://github.com/processone/ejabberd-contrib) 

I've implemented that in commit 
https://github.com/dirkmoors/docker-ejabberd/commit/13bfa344733753e9f0a6978d5b508e2d8b63140f

dirkmoors commented:
This looks promising to integrate the clustering as a POST command?

http://easyerl.blogspot.be/2012/09/easy-ejabberd-clustering-procedure.html

rroemhild commented:
Just four more things I would like to see in this PR:

1. I don't think that it's necessary to change the module path via environment. Serving the static path `~/.ejabberd-modules` should be fine enough. Maybe ~/modules and then add it to VOLUMES.
2. Create the module path should be done in the install RUN section to omit an extra step.
3. If `erlc` fails to compile the build process stops and never print the error message, that's fine if something fails on build the image is not complete.
4. Compile any erl file in `scripts/erl`

Here is my diff how I imagine this feature:

```diff
diff --git a/Dockerfile b/Dockerfile
index 510f1f9..ebbbec8 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -6,7 +6,6 @@ ENV EJABBERD_USER ejabberd
 ENV EJABBERD_ROOT /opt/ejabberd
 ENV EJABBERD_WEB_ADMIN_SSL true
 ENV EJABBERD_S2S_SSL true
-ENV EJABBERD_CONTRIB_MODULES_PATH $EJABBERD_ROOT/.ejabberd-modules/
 ENV HOME $EJABBERD_ROOT
 ENV PATH $EJABBERD_ROOT/bin:/usr/sbin:/usr/bin:/sbin:/bin
 ENV DEBIAN_FRONTEND noninteractive
@@ -45,6 +44,7 @@ RUN gpg --keyserver hkp://hkps.pool.sks-keyservers.net --recv-keys 31468D18DF984
             --adminpw ejabberd \
     && rm -rf /tmp/* \
     && mkdir $EJABBERD_ROOT/ssl \
+    && mkdir $EJABBERD_ROOT/.ejabberd-modules \
     && rm -rf $EJABBERD_ROOT/database/ejabberd@localhost
 
 # Make config
@@ -60,14 +60,9 @@ COPY ./run $EJABBERD_ROOT/bin/run
 # Add run scripts
 ADD ./scripts $EJABBERD_ROOT/bin/scripts
 
-# Prepare output directory for plugins
-RUN mkdir -p $EJABBERD_CONTRIB_MODULES_PATH
-
-# Compile easy cluster. The output files have to be in the /lib/ejabberd/ebin/ directory!
-RUN erlc -W0 -o $EJABBERD_CONTRIB_MODULES_PATH $EJABBERD_ROOT/bin/scripts/erl/easy_cluster.erl
-
-# Check for compiled plugins
-RUN if [ ! -f $EJABBERD_CONTRIB_MODULES_PATH/easy_cluster.beam ]; then echo "Failed to compile erlang plugin!"; exit 1; fi
+# Compile erl scripts
+RUN erlc -W0 -o $EJABBERD_ROOT/.ejabberd-modules \
+    $EJABBERD_ROOT/bin/scripts/erl/*.erl
 
 VOLUME ["$EJABBERD_ROOT/database", "$EJABBERD_ROOT/ssl"]
 EXPOSE 5222 5269 5280 4560
diff --git a/ejabberdctl.cfg.tpl b/ejabberdctl.cfg.tpl
index 1431c98..6fa1641 100644
--- a/ejabberdctl.cfg.tpl
+++ b/ejabberdctl.cfg.tpl
@@ -158,10 +158,11 @@ EJABBERD_BYPASS_WARNINGS=true
 
 
 #.
-#' CONTRIB_MODULES_PATH: Path to directory where extra modules (*.beam) are installed
+#' CONTRIB_MODULES_PATH:
+
+# Path to directory where extra modules (*.beam) are installed.
 #
 # Source: https://github.com/processone/ejabberd-contrib
 #
 # Default: ~/.ejabberd-modules
-CONTRIB_MODULES_PATH={{ env['EJABBERD_CONTRIB_MODULES_PATH'] or "~/.ejabberd-modules" }}
-
+CONTRIB_MODULES_PATH="/opt/ejabberd/.ejabberd-modules"
[rroe:~/Workspace/docker/rroemhild/ejabberd] auto-cluster(+8/-12)* ± 
[rroe:~/Workspace/docker/rroemhild/ejabberd] auto-cluster(+11/-12)* ± git diff
diff --git a/Dockerfile b/Dockerfile
index 510f1f9..cc2f732 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -6,7 +6,6 @@ ENV EJABBERD_USER ejabberd
 ENV EJABBERD_ROOT /opt/ejabberd
 ENV EJABBERD_WEB_ADMIN_SSL true
 ENV EJABBERD_S2S_SSL true
-ENV EJABBERD_CONTRIB_MODULES_PATH $EJABBERD_ROOT/.ejabberd-modules/
 ENV HOME $EJABBERD_ROOT
 ENV PATH $EJABBERD_ROOT/bin:/usr/sbin:/usr/bin:/sbin:/bin
 ENV DEBIAN_FRONTEND noninteractive
@@ -45,6 +44,7 @@ RUN gpg --keyserver hkp://hkps.pool.sks-keyservers.net --recv-keys 31468D18DF984
             --adminpw ejabberd \
     && rm -rf /tmp/* \
     && mkdir $EJABBERD_ROOT/ssl \
+    && mkdir $EJABBERD_ROOT/modules \
     && rm -rf $EJABBERD_ROOT/database/ejabberd@localhost
 
 # Make config
@@ -60,16 +60,14 @@ COPY ./run $EJABBERD_ROOT/bin/run
 # Add run scripts
 ADD ./scripts $EJABBERD_ROOT/bin/scripts
 
-# Prepare output directory for plugins
-RUN mkdir -p $EJABBERD_CONTRIB_MODULES_PATH
+# Compile erl scripts
+RUN erlc -W0 -o $EJABBERD_ROOT/modules \
+    $EJABBERD_ROOT/bin/scripts/erl/*.erl
 
-# Compile easy cluster. The output files have to be in the /lib/ejabberd/ebin/ directory!
-RUN erlc -W0 -o $EJABBERD_CONTRIB_MODULES_PATH $EJABBERD_ROOT/bin/scripts/erl/easy_cluster.erl
+VOLUME ["$EJABBERD_ROOT/database",
+        "$EJABBERD_ROOT/ssl",
+        "$EJABBERD_ROOT/modules"]
 
-# Check for compiled plugins
-RUN if [ ! -f $EJABBERD_CONTRIB_MODULES_PATH/easy_cluster.beam ]; then echo "Failed to compile erlang plugin!"; exit 1; fi
-
-VOLUME ["$EJABBERD_ROOT/database", "$EJABBERD_ROOT/ssl"]
 EXPOSE 5222 5269 5280 4560
 
 CMD ["start"]
diff --git a/ejabberdctl.cfg.tpl b/ejabberdctl.cfg.tpl
index 1431c98..6fa1641 100644
--- a/ejabberdctl.cfg.tpl
+++ b/ejabberdctl.cfg.tpl
@@ -158,10 +158,11 @@ EJABBERD_BYPASS_WARNINGS=true
 
 
 #.
-#' CONTRIB_MODULES_PATH: Path to directory where extra modules (*.beam) are installed
+#' CONTRIB_MODULES_PATH:
+
+# Path to directory where extra modules (*.beam) are installed.
 #
 # Source: https://github.com/processone/ejabberd-contrib
 #
 # Default: ~/.ejabberd-modules
-CONTRIB_MODULES_PATH={{ env['EJABBERD_CONTRIB_MODULES_PATH'] or "~/.ejabberd-modules" }}
-
+CONTRIB_MODULES_PATH="/opt/ejabberd/modules"
```

rroemhild commented:
Hmmm ejabberdctl is not able to load the module.

```
(ejabberd@localhost)1> l('easy_cluster').
{error,nofile}
```

rroemhild commented:
`/opt/ejabberd/lib/ejabberd-15.03/ebin` seems to be the right directory for this plugin.

rroemhild commented:
So `CONTRIB_MODULES_PATH` is a good thing to add but not necessary for the cluster plugin. I would like to see a new clear PR for this feature to the master branch without the module parts and I'll add the module part later if this is ok for you.

dirkmoors commented:
Ok, I'll look into it this evening.

rroemhild commented:
With [15.03](https://blog.process-one.net/ejabberd-15-03/) clustering was improved. Did you test this? http://docs.ejabberd.im/admin/guide/clustering/

dirkmoors commented:
No I didn't know this...I'll give it a go, definitely!

dirkmoors commented:
I've renamed the branch: https://github.com/dirkmoors/docker-ejabberd/tree/ssl-fixes
